consensus/consensus_v2.go

```diff
@@ -1073,33 +1094,79 @@ func (consensus *Consensus) Start(blockChannel chan *types.Block, stopChan chan
 				consensus.getLogger().Info().
 					Uint64("MsgBlockNum", newBlock.NumberU64()).
 					Msg("[ConsensusMainLoop] Received Proposed New Block!")
+
+				//VRF/VDF is only generated in the beacon chain
 				if consensus.ShardID == 0 {
-					// TODO ek/rj - re-enable this after fixing DRand
-					//if core.IsEpochBlock(newBlock) { // Only beacon chain do randomness generation
-					//	// Receive pRnd from DRG protocol
-					//	consensus.getLogger().Debug("[DRG] Waiting for pRnd")
-					//	pRndAndBitmap := <-consensus.PRndChannel
-					//	consensus.getLogger().Debug("[DRG] Got pRnd", "pRnd", pRndAndBitmap)
-					//	pRnd := [32]byte{}
-					//	copy(pRnd[:], pRndAndBitmap[:32])
-					//	bitmap := pRndAndBitmap[32:]
-					//	vrfBitmap, _ := bls_cosi.NewMask(consensus.PublicKeys, consensus.leader.ConsensusPubKey)
-					//	vrfBitmap.SetMask(bitmap)
-					//
-					//	// TODO: check validity of pRnd
-					//	newBlock.AddVrf(pRnd)
-					//}
-
-					rnd, blockHash, err := consensus.GetNextRnd()
+					// generate VRF if the current block has a new leader
+					if !consensus.ChainReader.IsSameLeaderAsPreviousBlock(newBlock) {
+						vrfBlockNumbers, err := consensus.ChainReader.ReadEpochVrfBlockNums(newBlock.Header().Epoch)
+						if err != nil {
+							consensus.getLogger().Error().Err(err).
+								Uint64("MsgBlockNum", newBlock.NumberU64()).
+								Msg("[ConsensusMainLoop] failed to read VRF block number from local db")
```

Comment:
return here if err != nil ?

Reply:
here the err just means there's no VRF numbers cached for this epoch.  so we need to generate new ones by continuing the execution below.  should I use info instead of error for the message ?

Reply:
Got it, in this case probably info is better.